grabber/main.go

```diff
@@ -208,77 +217,103 @@ func checkAncestors(importer *backend.Backend, blockNumber int64, generations in
+	for blockNumber > oldest && importer.NeedReloadParent(blockNumber) {
+		lgr := importer.Lgr.With(zap.Int64("blockNumber", blockNumber))
+		lgr.Info("Redownloading corrupted or missing ancestor")
 		block, err := importer.BlockByNumber(blockNumber)
 		if err != nil {
-			importer.Lgr.Fatal("blockByNumber - checkAncestors", zap.Int64("blockNumber", blockNumber), zap.Error(err))
-		}
-		if block != nil {
-			importer.ImportBlock(block)
-			if err != nil {
-				importer.Lgr.Fatal("importBlock - checkAncestors", zap.Error(err))
-			}
+			lgr.Error("Failed to get ancestor", zap.Error(err))
+			time.Sleep(5 * time.Second)
+			continue
+		} else if block == nil {
+			lgr.Error("Ancestor not found")
+			time.Sleep(5 * time.Second)
+			continue
 		}
+		importer.ImportBlock(block)
+		blockNumber--
 	}
 }
 
 // checkTransactionsConsistency checks if the block tx count matches the number of txs, and reimports the block if not.
 // If a block is reimported, then the parent hash is returned.
-func checkTransactionsConsistency(importer *backend.Backend, blockNumber int64) common.Hash {
+func checkTransactionsConsistency(importer *backend.Backend, blockNumber int64) (common.Hash, error) {
 	if !importer.TransactionsConsistent(blockNumber) {
 		importer.Lgr.Info("Redownloading block because number of transactions are wrong", zap.Int64("blockNumber", blockNumber))
 		block, err := importer.BlockByNumber(blockNumber)
 		if err != nil {
-			importer.Lgr.Fatal("checkTransactionsConsistency", zap.Int64("blockNumber", blockNumber), zap.Error(err))
+			return common.Hash{}, fmt.Errorf("failed to get block: %v", err)
 		}
 		if block != nil {
 			importer.ImportBlock(block)
-			return block.ParentHash()
+			return block.ParentHash(), nil
 		}
 	}
-	return common.Hash{}
+	return common.Hash{}, nil
 }
 
 func updateAddresses(sleep time.Duration, updateContracts bool, blockRangeLimit uint64, workersCount uint, importer *backend.Backend) {
 	lastUpdatedAt := time.Unix(0, 0)
 	lastBlockUpdatedAt := int64(0)
+	t := time.NewTicker(sleep)
+	defer t.Stop()
 	for {
-		start := time.Now()
-		currentTime := time.Now()
-		currentBlock := getLatestBlockNumber(importer)
-		addresses := importer.GetActiveAdresses(lastUpdatedAt, updateContracts)
-		importer.Lgr.Info("updateAddresses:", zap.Time("lastUpdatedAt", lastUpdatedAt), zap.Int("Addresses in db", len(addresses)))
-		var jobs = make(chan *models.ActiveAddress, workersCount)
-		var wg sync.WaitGroup
-		for i := 0; i < int(workersCount); i++ {
-			wg.Add(1)
-			go worker(&wg, jobs, currentBlock, blockRangeLimit, importer)
-		}
-		for _, address := range addresses {
-			jobs <- address
+		select {
+		case <-t.C:
+			start := time.Now()
+			currentTime := time.Now()
+			currentBlock, err := importer.GetLatestBlockNumber()
+			if err != nil {
+				importer.Lgr.Error("Update Addresses: Failed to get latest block number", zap.Error(err))
+				continue
```

Comment:
Some of the fatals here will end the execution, which is effectively 'retried' on the next tick.